source/common/protobuf/message_converter_utility.cc

```diff
@@ -0,0 +1,127 @@
+#include "source/common/protobuf/message_converter_utility.h"
+
+#include <limits>
+#include <memory>
+#include <string>
+
+#include "source/common/common/logger.h"
+
+#include "absl/log/absl_check.h"
+#include "grpc_transcoding/message_reader.h"
+
+namespace Envoy::ProtobufMessage {
+
+namespace {
+using ::google::grpc::transcoding::kGrpcDelimiterByteSize;
+using ::google::grpc::transcoding::MessageAndGrpcFrame;
+using ::google::grpc::transcoding::MessageReader;
+} // namespace
+
+absl::StatusOr<MaybeParseGrpcMessageOutput>
+MaybeParseGrpcMessage(CreateMessageDataFunc& factory, Envoy::Buffer::Instance& request_in) {
+  // Parse the gRPC frame header.
+  if (request_in.length() < kGrpcDelimiterByteSize) {
+    ENVOY_LOG_MISC(info, "Need more data for gRPC frame header parsing. Current size={}",
+                   request_in.length());
+    MaybeParseGrpcMessageOutput ret;
+    ret.needs_more_data = true;
+    return ret;
+  }
+  unsigned char frame_header[kGrpcDelimiterByteSize];
+  request_in.copyOut(0, kGrpcDelimiterByteSize, &frame_header);
+
+  // Parse the message size.
+  auto message_size_status = DelimiterToSize(frame_header);
+  if (!message_size_status.ok()) {
+    return message_size_status.status();
+  }
+  const uint64_t message_size = message_size_status.value();
+  const uint64_t total_size = message_size + kGrpcDelimiterByteSize;
+  if (request_in.length() < total_size) {
+    ENVOY_LOG_MISC(info, "Need more data for gRPC frame. Current size={}, needed={}",
+                   request_in.length(), total_size);
+    MaybeParseGrpcMessageOutput ret;
+    ret.needs_more_data = true;
+    return ret;
+  }
+
+  // Drain out the gRPC frame header as we already buffered it separately.
+  MaybeParseGrpcMessageOutput output;
+  output.frame_header.move(request_in, kGrpcDelimiterByteSize);
+
+  // Move the required data to its own BEFORE any parsing occurs.
+  // Data cannot be moved again after parsing.
+  output.owned_bytes = std::make_unique<Envoy::Buffer::OwnedImpl>();
+  output.owned_bytes->move(request_in, message_size);
+
+  // Construct the RawMessage for extraction.
+  // Get enough data from the buffer for the first gRPC message.
+  uint64_t bytes_needed = message_size;
+  output.message = factory();
+  const Envoy::Buffer::RawSliceVector slices = output.owned_bytes->getRawSlices();
+  for (auto slice : slices) {
```

Comment:
I'm not quite sure what type `slice` is, but should we consider making this const&? or just &?